### DISCUSSION TITLE
More Meta-Data on installers

### DISCUSSION THREAD

**Discussion by denelon** [microsoft/winget-cli#615] [Ideas]
Hello Everyone,

I wanted to get some feedback for additional meta-data on installers. We've been observing several scenarios related to different installers, and I was hoping for some lively discussion. We're thinking about a more rich manifest to help with different scenarios:

1. We've seen several cases where installers require elevation.
This is likely to be handled in #149 
2. Some installers require the acceptance of a EULA.
3. Some installers require the user to select a language, or other user "interaction".
4. " " " restart.

What other kinds of meta-data should we consider capturing in the manifest to enable improvements in the UX with the client?

For example. If we knew which installers needed EULA from the user, we could install those first when we're doing import/export so you could accept them, and head off to grab a cup of coffee (or any other favorite beverage) while the installation continued.

**Comment by zakius** [comment]
Personally I think we should avoid *installers* as much as possible and instead use packages and metapackages telling winget what to do. If you want to install only some features of an app there's no point in pulling whole installer. And when you want to add (or even remove!) new ones why would you pull the whole installer again? Or cache it all the time? But that's the ideal world, where distributors want to cooperate and provide proper packages, much more efficient in technological means but will take years before we get even close to having proper package repository so supplementing it with installers is inevitable.

I think most of things can be worked around with metapackages pulling the same installer and running it with different parameters, though it'd have to be *smart* enough to run it once and install all features at once instead of trying to do that multiple times in case of feature selection. For EULA there's always auto answering. The biggest issue comes from installers that don't have unattended option available and in my opinion these shouldn't be available in repository at all.

**Comment by MathiasMagnus** [comment]
Not sure if this is something that can be captured in the sense it was asked, but installers can provide options which only light up in the presence of another package installed (Visual Studio / MSBuild integration for eg.). Would it be possible to have packages install dependents automatically when such options are requested? (I'm feeling that hell starts breaking loose quickly; unsure if it's possible to wrap all kinds of installers without true repackaging.)

**Reply by denelon** [reply]
@mathiasMagnus, can you give an example of one or two of these? We've got the work for "explicit" dependencies coming in #163, but this sounds a bit different.

**Reply by MathiasMagnus** [reply]
The installer of Codeplay's [ComputeCpp](https://developer.codeplay.com/products/computecpp/ce/home) (a compiler implementing the SYCL, a GPGPU API) has options in the installer that light up only when VS is installed and allows integrating into VS by installing MSBuild schemas that make VS SYCL aware by adding the extra build steps required for MSVC and `compute++.exe` to compile a SYCL program. Very same thing is true for the [oneAPI Base Kit](https://software.intel.com/content/www/us/en/develop/tools/oneapi/base-toolkit.html), which is Intel's implementation of the same standard (and more). It too integrates with VS, but the options only light up if it detects VS as being present.

Surely there are other examples of applications with plug-in systems that 3rd party tools integrate with. (I could also say nSight from Nvidia, which again has a standalone app but can also integrate into VS or the (now discontinued) CodeXL, a similar GPU debugger / profiler from AMD, which had a standalone app, but could also integrate into VS.) If I were hardcore Houdini, Maya, Photoshop etc. user I may be able to name similar applications which integrate into other ones, but only when peresent.

**Reply by denelon** [reply]
Interesting. It sounds like it's not really a hard dependency, but more of an integration. "IntegratesWith: Microsoft.VisualStudioCode" could be some kind of a hint to the CLI. Does the order of installation matter?

**Reply by MathiasMagnus** [reply]
These examples all rely on VS being installed before, as that provides the install location of the integration component. Without repackaging, the plug-in likely needs to be installed after the base SW. With proper repackaging it may be the other way around, but that's sort of like eating the cake before it's even baked.

This "Integrates with" may not fundamentally differ from an extension which is provided by a 3rd party. Much like there were Edge extensions in the MS Store provided by 3rd parties which enhance the browser (and may have entry points of their own).

**Reply by denelon** [reply]
Do you think "Extends" would make better sense? The client will be aware of what packages it has installed, and in many cases even packages that were installed via Add/Remove programs from the Control Panel as part of the "List" feature. We could likely check to see if VS is already installed and provide some kind of message to the user.

If the other package has not been installed:

`winget install <some extension>`
`<some extension> has been identified as an extension of <package>.`
`Would you like to install <package> first [Y/N]`

Are there any cases of extensions that support more than one "base" package? I'm curious about having a list of them.

`winget install <some extension>`
`<some extension> has been identified as an extension of <package a>, <package b>, and <package c>.`
`Do you wish to install <some extension> [Y/N]`

I think we may also want to have a force flag to avoid the prompt/interruption in flow. What are your thoughts on usability?

**Reply by MathiasMagnus** [reply]
_(Thinking out loud.)_

The root cause of the guessing game here is that some installers bundle multiple packages into one installer (in this case, a compiler and build system extensions to make life easier to some). Guessing / prompting comes in the form of: "dear user, are you interested in said (sub) feature?" Both ComputeCpp and oneAPI are good examples that they have multiple "base" packages, as they both provide VS2017 and VS2019 integration. Hell I don't want both of them installed. Because I'm using both SDKs with CMake, I don't need VS integration, even if it's just a few XML files on disk, so I wouldn't be offended if it were installed without me asking for it. Both these SDKs work flawlessly with Clang and CMake, so by no means are these extensions necessary at all, it really is an optional feature of them which ideally should live in different packages which hard depend on the MSBuild component. (Not even VS or just the Build Tools, but a specific component from either.) Note how I mentioned Qt as a dual-licensed product, but it too has lot of components which depend on specific VS workloads / components.

As for "what could we do today?" both prompts look okay to me. The second case may also be:
```
winget install <some extension>
<some extension> has been identified as an extension of <package a>, <package b>, and <package c>.
List packages that should be installed previously to <some extension> (leave empty for none): [name1, name2...]
```

**Reply by jedieaston** [reply]
Prior art of a feature like this exists in Scoop, they call it "suggests" in the manifest: https://github.com/lukesampson/scoop/wiki/App-Manifests.

**Comment by jedieaston** [comment]
A easy one to add is something like "nonCommercialLicense" to flag packages that shouldn't be installed in corporate environments due to licensing issues (like WinRAR), which would then make the client warn the person installing (if a setting was checked, perhaps via GPO so that if users are allowed to install from the community repo, they can't install packages that they would be in violation of the EULA for).

**Reply by denelon** [reply]
@jedieaston, that's an interesting case. I read the license at https://www.win-rar.com/winrarlicense.html. It looks like it requires a license 40 days after trial. Are you aware of any other packages like this? I can see how that would be very troubling for enterprise scenarios.

**Reply by MathiasMagnus** [reply]
[Qt](https://www.qt.io/download) is the prime example here which has both commercial and non-commercial licenses. The web-site is pretty vocal about the commercial version, but open-source devs may obtain the OSS SDK if they develop applications with compatible licensing.

**Reply by jedieaston** [reply]
@denelon I recently updated [XnViewClassic](https://www.xnview.com/en/xnview/) which is specifically licensed for non-commercial use (companies must purchase a license), but the application still starts and someone could not realize the license until it's too late.

Another example is Visual Studio 2019 Community, which [does not allow enterprise use](https://visualstudio.microsoft.com/license-terms/mlt031819/) unless you are using it for open-source or Windows driver development (which would both be good reasons to have an override flag in the winget client on a per-case basis if you had the setting activated to block "non commercially licensed" software).

**Reply by denelon** [reply]
OK, so something like "NonCommercialLicense: true" in the manifest would identify these kinds of packages. Some kind of flag in the settings like "WarnOnNonCommercialLicense: true" could trigger behavior like provided below.

`winget install <package>`
`Warning <package> has a non-commercial license.`
`Do you wish to install <package> [Y/N]?`

If this were enforced by Group Policy, the user would not get the warning and would not be able to continue/override.

If the user were to pass a "force" flag, the warning would not be rendered.

Is this what you are thinking?

**Reply by jedieaston** [reply]
That's exactly what I'm thinking. It would be insanely handy.

**Comment by sharpninja** [comment]
I think making it easy to install Visual Studio itself would be huge.  Supply a comma separated list of features on the command line as arguments or the location of a config file (which could be in a Gist or some other place on the web).

**Reply by denelon** [reply]
@sharpninja you can use the --override option during install. Is that what you were looking for?

The options for install are listed in the help:
`winget install -?`

**Reply by sharpninja** [reply]
What options would you choose?  Can you give an example of installing 2019 community preview with desktop development features?

**Comment by jedieaston** [comment]
Something else I thought of: certificates for drivers. If an application installs a driver, we should have a way to have the certificate for the driver in the manifest (available via a URL probably), so that winget could install the certificate first so that the driver could be installed silently. I probably will write a formal feature request issue for this to explain the flow I'm thinking of, but for things like PlayStation Now (which needs to install a game controller driver) or VPN clients that need to install a driver, this is necessary to make sure they can install silently.

In Chocolatey (which is obviously a completely different ball game since they are running PowerShell scripts), they solved this by downloading the certificate and running certutil prior to executing the installer, see https://chocolatey.org/packages/nextdns for an example.

**Reply by denelon** [reply]
If the driver has an installer, it could be identified as a dependency (once dependency support is enabled and a package exists in the same source). If it's just an .exe or a .zip we'll need support for that as well. I'm not sure about how we would tackle installing certificates with the Windows Package Manager. That's a big security concern we'd have to deep dive into. It's also something that IT pros would very likely want to manage with Group Policy.

**Reply by jedieaston** [reply]
The problem with identifying them as dependencies is that often a driver is embedded in a installer, and Windows pops up a dialog box asking if the user wants to allow installing the driver. If you already trust the certificate the driver is signed with, this box is suppressed. Of course, if winget ends up being able to trust certificates (even if only in certain stores), I concur that GPO will really be needed to control this behavior.

**Reply by denelon** [reply]
Thanks for the clarity! Do you have an example for one of these? I'd like to meet with some other teams here at Microsoft to start the discussion on what this might look like.

**Reply by jedieaston** [reply]
Absolutely, NextDNS is a good example of what I am talking about. If you run ` winget install NextDNS `, during the install process, the NextDNS installer will try to install the OpenVPN TAP driver, resulting in this:

![image](https://user-images.githubusercontent.com/21368066/111207819-79396c00-85a0-11eb-8e46-2ddd4c67f37c.png)


However, if you import OpenVPN's driver signing certificate into the TrustedPublishers store first (I simulated this by exporting the certificate using certmgr.msc, opening a new sandbox and re-importing the certificate using certutil) before installing NextDNS, that dialog box does not appear, and NextDNS installs silently as expected.

**Reply by jedieaston** [reply]
Another example is Ultimaker.Cura, it doesn't install anymore due to wanting to install drivers: microsoft/winget-pkgs#19798.

**Comment by jedieaston** [comment]
I don't know if this is a discussion for now or for a ` ManifestSpec: 1.1 `, but one thing I thought about when reviewing the new specification and writing automation for it was specific command line switches depending on locale. [Visual Studio](https://docs.microsoft.com/en-us/visualstudio/install/use-command-line-parameters-to-install-visual-studio?view=vs-2019), for example, only installs additional locales based on ` --addProductLang <language-locale> `, (assuming that the user isn't running Windows in the locale they want for VS, which happens more often than not) and I'm sure other applications do too. 

Is this something worth worrying about at this stage? It could cause unexpected behavior if a user asks winget to install "es-MX" and it doesn't actually happen.

**Reply by denelon** [reply]
I think it is more of a client issue. The idea for now is that you could have meta-data that is appropriately localized like "shortDescription". We've also encountered some packages that have an installer per locale. We will attempt to honor the Windows locale for displaying meta-data and selecting installers when one is present matching the user's Windows configuration. There might be a feature request need on an Issue for "Add an argument to specify installer locale".

We've been doing some thinking about preferences and requirements in this context. If I prefer something and it's available I expect the behavior to match my preference. If I require something and it is not available, then I will be informed my requirement is not available. Or in some use cases, my requirement might reduce the results returned during search or some other commands.
